restutil/processor.go

```diff
@@ -0,0 +1,707 @@
+package restutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+	"gopkg.in/cheggaaa/pb.v1"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	Transport = &http.Transport{
+		MaxIdleConnsPerHost: 128,
+		Dial: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).Dial,
+	}
+
+	HttpClient = &http.Client{
+		Transport: Transport,
+	}
+)
+
+type binaryMsg struct {
+	id   *string
+	body *io.ReadCloser
+}
+
+func PutAllBinaryRest(baseFromURL string, baseToURL string, user string, pass string, conns int, throttle int, dumpFailed bool) (err error) {
+	msgs := make(chan *binaryMsg, 128)
+	var failChan chan []byte
+	rp := &resourcePutter{
+		baseURL: baseToURL,
+		user:    user,
+		pass:    pass,
+	}
+
+	errs := make(chan error)
+	failwg := sync.WaitGroup{}
+	if dumpFailed {
+		failChan = make(chan []byte, conns*2)
+		failwg.Add(1)
+		go func() {
+			defer failwg.Done()
+
+			for resource := range failChan {
+				_, err := os.Stdout.Write(resource)
+				if err == nil {
+					_, err = io.WriteString(os.Stdout, "\n")
+				}
+				if err != nil {
+					select {
+					case errs <- err:
+					default:
+					}
+					return
+				}
+			}
+		}()
+
+	}
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < conns; i++ {
+		wg.Add(1)
+		go rp.putAllBinary(msgs, failChan, &wg)
+	}
+	getAllBinary(baseFromURL, throttle, conns, msgs)
+	wg.Wait()
+
+	if dumpFailed {
+		close(failChan)
+		failwg.Wait()
+	}
+
+	select {
+	case err := <-errs:
+		return err
+	default:
+		return nil
+	}
+}
+
+func getAllBinary(baseURL string, throttle int, conns int, msgs chan *binaryMsg) (err error) {
+	ids := make(chan *string, conns*24)
```

Comment:
I rolled a die and that was the random generated. Wanted to make sure there was a nice buffer size of IDs that would grow with the number of concurrent connections we are using.  I should probably move it to a constant so it is more obvious rather than a magic number.